`General`: Add filter for exercises

florian-glombik opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.


### Motivation and Context
* Give students the option to filter exercises _(e.g. for exam preparation)_


### Description
* Added filter options for exercises:
   - exercise categories
   - exercise type
   - difficulty
   - achieved score
   - achievable points
* The filters are considering general present values _(e.g. if there was no Quiz exercise in a course, the exercise type `quiz` will not be an available filter option; the same applies to the general filter option - if there is only one exercise type in the course the filter option will not be displayed)_
* Introduced a shared component for the `CustomExerciseCategoryBadgeComponent`
* Adjusted the usage of ExerciseCategory -> now really using the class object (and using equals and compare function) instead of actually having a string object in the background
* Added a `RangeSlider` component as the `ngBootstrap` does not seem to offer such a functionality (https://getbootstrap.com/docs/5.0/forms/range/) 

### Steps for Testing
<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
Prerequisites:
- 1 Student _(participating in a course with many exercises)_

1. Navigate to a course with many exercises
2. Open the filter Modal and apply the filters in different combinations _(state the combinations in your review comment)_
3. Verify that the filter works according to your expectations
4. Verify that only filter options are displayed that differentiate the exercises


### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked
> Click on the badges to get to the test servers.

[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)](https://artemis-test1.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)](https://artemis-test2.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)](https://artemis-test3.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)](https://artemis-test4.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)](https://artemis-test5.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)](https://artemis-test6.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)](https://artemis-test9.artemis.cit.tum.de)

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2


### Test Coverage
#### Client

| Class/File | Line Coverage |                                 Confirmation (assert/expect) |
|------------|--------------:|------------------------------------------------------------:|
| exercise-category.model.ts | 100% |                                                            ✅ |
| custom-exercise-category-badge.component.ts | 55.55% |  ❌ no reasonable logic, introduced for reusability of badges |
| exercise-filter-modal.component.ts | 88% |                                                          ✅ |
| exercise-filter-modal.helper.ts | 88.88% |                                                          ✅ |
| range-slider.component.ts | 71.87% |                                                          ✅ |
| sidebar.component.ts | 87.69% |                                                          ✅ |
| sidebar.helper.ts | 100% |                                                            ✅ |
| utils.ts | 96.61% |                                                          ✅  |

### Screenshots
![image](https://github.com/ls1intum/Artemis/assets/63976129/a74cdf6a-8149-4e7a-8c50-154696ac51c9)

![image](https://github.com/ls1intum/Artemis/assets/63976129/4f6b9404-6b7d-4f7f-9187-beed64aed7d8)

![exampleOfFilterUsage](https://github.com/ls1intum/Artemis/assets/63976129/76ecdf41-fed2-4d82-8ff1-614cd3c08d69)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added a modal for filtering exercises based on categories, types, difficulty levels, achieved scores, and points in the student dashboard.
  - Introduced a badge component for visually representing exercise categories and enhancing user interaction.
  - Updated the student dashboard with new messages in German for clearer feedback on exercise filtering.
  - Streamlined exercise category display by implementing a custom badge component.

- **Enhancements**
  - Improved sidebar to dynamically display filter options based on the context.
  - Enhanced the appearance and functionality of dropdown menus in the filter modal.
  - Implemented better user interaction for the range slider with visual progress indicators.
  - Added a filter feature to the sidebar for exercise-related navigation.

- **Bug Fixes**
  - Ensured proper initialization and teardown of filter components to avoid memory leaks.

- **Tests**
  - Added unit tests for the range slider and exercise filter components to validate their behavior and event handling.
  - Enhanced testing framework integration for the sidebar component to thoroughly validate filtering functionality.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

florian-glombik commented on src/main/webapp/app/shared/chart/chart-category-filter.ts:
```diff
@@ -102,9 +102,7 @@ export class ChartCategoryFilter extends ChartFilter {
-            return exercise
-                .categories!.flatMap((category: ExerciseCategory) => this.filterMap.get(category.category!)!)
-                .reduce((value1: boolean, value2: boolean) => value1 || value2);
+            return exercise.categories!.flatMap((category: ExerciseCategory) => this.filterMap.get(category.name!)!).reduce((value1: boolean, value2: boolean) => value1 || value2);
```
No change in logic, this line is now only wrapped because it is a few chars shorter after renaming

florian-glombik commented on src/main/webapp/app/exercises/shared/exercise/exercise.service.ts:
```diff
@@ -362,12 +362,15 @@ export class ExerciseService {
     static parseExerciseCategories(exercise?: Exercise) {
         if (exercise?.categories) {
-            exercise.categories = exercise.categories.map((category) => JSON.parse(category as string) as ExerciseCategory);
+            exercise.categories = exercise.categories.map((category) => {
+                const categoryObj = JSON.parse(category as unknown as string);
+                return new ExerciseCategory(categoryObj.color, categoryObj.category);
+            });
```
We need to use a real object if we want to use the `equals` method

pzdr7 commented on src/main/webapp/app/entities/exercise-category.model.ts:
```diff
@@ -1,6 +1,30 @@
     public category?: string;
 
-    constructor() {}
+    constructor(color: string | undefined, category: string | undefined) {
```
Would this work?

```suggestion
    constructor(color?: string, category?: string) {
```

JohannesStoehr replied:
It would work, but would change the semantics of this method. `: string | undefined` requires a parameter that might also be undefined `?: string` is an optional parameter that may be undefined or can even be left out when calling this method

florian-glombik replied:
I defined it like that on purpose to prevent accidentally only the color is set and always bought parameters need to be defined - I just changed the order of the category and color in the constructor to make it less error-prone if it is changed in the future and the color is left out

I do slightly prefer the current approach

rabeatwork commented:
Here are some UI improvements I discovered while testing:
- Can you add a bit more spacing between the filter and the searchbar?
- We need a second state for the Filter in the sidebar that shows that we have some applied filter 
- Can you add a "Reset Filter" to the Modal Dialog? (Modal Footer Left Side)
- Can you reduce the size of the selected categories? 
- I understand why you removed the input field for the categories, but it's somewhat confusing. Could you instead add a non-selectable option that says "no more options" or something similar?
- And last but also least, could you please remove the margin/padding at the beginning of the checkboxes? If you're using a for loop and want to keep the margin for the second, third, and subsequent items, you can use $first to remove the spacing only for the first item.

![image](https://github.com/ls1intum/Artemis/assets/75392103/ee0ff3b0-4850-41ec-bd20-d229d4c76382)

![image](https://github.com/ls1intum/Artemis/assets/75392103/4f6b68ce-c26d-427d-867e-d7f3b85a414b)

![Screenshot 2024-07-05 at 16 04 28](https://github.com/ls1intum/Artemis/assets/75392103/46a3788d-6c67-4b1f-9141-119460bb6f68)

I will take a look at the code in the upcoming days.

pzdr7 commented on src/main/webapp/i18n/en/student-dashboard.json:
```diff
@@ -33,7 +33,8 @@
             "general": {
-                "noDataFound": "No data found."
+                "noDataFound": "No data found.",
+                "noExercisesFoundWithAppliedFilter": "No exercises were found to which the selected filters apply."
```
No strong opinion here - just an idea:

```suggestion
                "noExercisesFoundWithAppliedFilter": "No exercises match the selected filters."
```

JohannesWt commented on src/main/webapp/app/shared/category-selector/category-selector.component.ts:
```diff
@@ -134,10 +133,7 @@ export class CategorySelectorComponent implements OnChanges {
-        category.category = categoryString;
-        category.color = this.chooseRandomColor();
-        return category;
+        return new ExerciseCategory(this.chooseRandomColor(), categoryString);
```
Just an idea. But I think this method in general is not really needed as its just called in one place and just creates a new object.

florian-glombik replied:
It is actually used in two places, so I think the method is reasonable 😅

JohannesWt commented on src/main/webapp/app/shared/exercise-categories/custom-exercise-category-badge.component.ts:
```diff
@@ -0,0 +1,22 @@
+export class CustomExerciseCategoryBadgeComponent {
+    @Input() category: ExerciseCategory;
+    @Input() displayRemoveButton: boolean = false;
+    @Input() onClick: () => void = DEFAULT_ON_CLICK;
```
Is this not working as default? (not sure)
```suggestion
    @Input() onClick: () => void = () => {};
```

JohannesWt commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:
```diff
@@ -0,0 +1,194 @@
+import { DifficultyLevel, ExerciseType } from 'app/entities/exercise.model';
+import { ExerciseCategory } from 'app/entities/exercise-category.model';
+
+export type FilterDetails = {
+    searchedTypes: ExerciseType[] | undefined;
+    selectedCategories: ExerciseCategory[];
+    searchedDifficulties: DifficultyLevel[] | undefined;
+    isScoreFilterApplied: boolean;
+    isPointsFilterApplied: boolean;
+    achievedScore: RangeFilter | undefined;
+    achievablePoints: RangeFilter | undefined;
+};
```
Can we not define undefined variables with `?` like this:
```suggestion
export type FilterDetails = {
    searchedTypes?: ExerciseType[];
    selectedCategories: ExerciseCategory[];
    searchedDifficulties?: DifficultyLevel[];
    isScoreFilterApplied: boolean;
    isPointsFilterApplied: boolean;
    achievedScore?: RangeFilter;
    achievablePoints?: RangeFilter;
};
```

JohannesWt commented on src/main/webapp/app/shared/sidebar/sidebar.component.ts:
```diff
@@ -36,15 +46,20 @@ export class SidebarComponent implements OnDestroy, OnChanges, OnInit {
-    faMagnifyingGlass = faMagnifyingGlass;
-    faChevronRight = faChevronRight;
-    faFilter = faFilter;
+    private modalRef: NgbModalRef | null;
```
is null here used because of any conditions of NgbModal? (And not undefined or like the following)
```suggestion
    private modalRef?: NgbModalRef;
```

JohannesWt commented on src/main/webapp/i18n/de/student-dashboard.json:
```diff
@@ -33,7 +33,8 @@
             "general": {
-                "noDataFound": "Keine Einträge gefunden."
+                "noDataFound": "Keine Einträge gefunden.",
+                "noExercisesFoundWithAppliedFilter": "Es wurden keine Aufgaben gefunden auf die die gewählten Filtereigenschaften zutreffen."
```
I think `zu den gewählten Filtern` is the best option

JohannesWt reviewed (changes requested):
Code in general looks good - just a few small comments

az108 commented on src/test/javascript/spec/component/shared/sidebar/sidebar.helper.spec.ts:
```diff
@@ -0,0 +1,338 @@
+
+describe('getExerciseTypeFilterOptions', () => {
+    it('should return present exercise types and sort them properly (same order as instructor creation)', () => {
+        const sidebarData: SidebarData = {
```
Is it possible to generalize this initialization as it is used frequently? If not it is also fine, no strong opinion on this.

florian-glombik replied:
This is only the same object for one test in `getExerciseTypeFilterOptions` and `getExerciseDifficultyFilterOptions` - I think that it is a better approach to keep it separated to reduce the possibility of side effects when adjusting the tests

My opinion is not very strong here. Do you disagree with that reasoning?

az108 reviewed (commented):
Tested on TS1 works as expected, no comments if all mentioned above are considered, will approve after changes 👍

florian-glombik commented on src/test/javascript/spec/component/shared/sidebar/sidebar-accordion.component.spec.ts:
```diff
@@ -93,6 +93,18 @@ describe('SidebarAccordionComponent', () => {
+    it('should call expandAll when filter is active', () => {
+        jest.spyOn(component, 'expandAll');
+
+        component.isFilterActive = true;
+        component.ngOnChanges();
+
+        fixture.detectChanges();
+
+        expect(component.expandAll).toHaveBeenCalled();
```
You're right, `detectChanges` is not sufficient but directly calling `ngOnChanges` is

Added the expected number of calls 👍

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.ts:
```diff
@@ -0,0 +1,78 @@
+        this.selectedMinValueChange.emit(updatedMinValue);
+    }
+
+    onSelectedMaxValueChanged(event: any): void {
```
I could not infer the type

I investigated it a little, and there does not seem to be a better solution than using `Event` as the type and casting to `HTMLInputElement` when needed

florian-glombik commented on src/main/webapp/app/shared/sidebar/sidebar.helper.ts:
What do you think about the added comments?

florian-glombik commented on src/main/webapp/app/shared/sidebar/sidebar.component.html:
```diff
@@ -4,12 +4,22 @@
                 <div
                     [ngClass]="{ 'content-height-dev': !isProduction || isTestServer }"
-                    jhiTranslate="artemisApp.courseOverview.general.noDataFound"
+                    [jhiTranslate]="
+                        (sidebarDataBeforeFiltering?.ungroupedData?.length ?? 0) > 0 && sidebarData.ungroupedData?.length === 0
+                            ? 'artemisApp.courseOverview.general.noExercisesFoundWithAppliedFilter'
+                            : 'artemisApp.courseOverview.general.noDataFound'
+                    "
```
They do exist

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:
```diff
@@ -0,0 +1,218 @@
+        this.activeModal.close();
+    }
+
+    search: OperatorFunction<string, readonly ExerciseCategoryFilterOption[]> = (text$: Observable<string>) => {
+        const debouncedText$ = text$.pipe(debounceTime(200), distinctUntilChanged());
+        const clicksWithClosedPopup$ = this.click$.pipe(filter(() => !this.instance.isPopupOpen()));
+        const inputFocus$ = this.focus$;
+
+        return merge(debouncedText$, inputFocus$, clicksWithClosedPopup$).pipe(
+            map((term) =>
+                term === ''
+                    ? this.selectableCategoryOptions
+                    : this.selectableCategoryOptions.filter(
+                          (categoryFilter: ExerciseCategoryFilterOption) => categoryFilter.category.category!.toLowerCase().indexOf(term.toLowerCase()) > -1,
+                      ),
+            ),
+        );
+    };
```
Does not work like that and will always be defined

florian-glombik commented on src/main/webapp/app/shared/sidebar/sidebar.component.html:
```diff
@@ -4,12 +4,28 @@
                 <div
                     [ngClass]="{ 'content-height-dev': !isProduction || isTestServer }"
-                    jhiTranslate="artemisApp.courseOverview.general.noDataFound"
+                    [jhiTranslate]="
+                        (sidebarDataBeforeFiltering?.ungroupedData?.length ?? 0) > 0 && sidebarData.ungroupedData?.length === 0
+                            ? 'artemisApp.courseOverview.general.noExercisesFoundWithAppliedFilter'
+                            : 'artemisApp.courseOverview.general.noDataFound'
+                    "
```
No, they are not missing - please do not suggest this again

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:
```diff
@@ -0,0 +1,208 @@
+        this.activeModal.close();
+    }
+
+    search: OperatorFunction<string, readonly ExerciseCategoryFilterOption[]> = (text$: Observable<string>) => {
+        const debouncedText$ = text$.pipe(debounceTime(200), distinctUntilChanged());
+        const clicksWithClosedPopup$ = this.click$.pipe(filter(() => !this.instance.isPopupOpen()));
+        const inputFocus$ = this.focus$;
+
+        return merge(debouncedText$, inputFocus$, clicksWithClosedPopup$).pipe(
+            map((term) =>
+                term === ''
+                    ? this.selectableCategoryOptions
+                    : this.selectableCategoryOptions.filter(
+                          (categoryFilter: ExerciseCategoryFilterOption) => categoryFilter.category.category!.toLowerCase().indexOf(term.toLowerCase()) > -1,
+                      ),
+            ),
+        );
+    };
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:
```diff
@@ -0,0 +1,208 @@
+        );
+    }
+
+    resetFilter() {
+        this.categoryFilter?.options.forEach((categoryOption) => (categoryOption.searched = false));
+        this.typeFilter?.options.forEach((typeOption) => (typeOption.checked = false));
+        this.difficultyFilter?.options.forEach((difficultyOption) => (difficultyOption.checked = false));
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:
```diff
@@ -0,0 +1,208 @@
+        return this.categoryFilter?.options.filter((categoryFilter) => !categoryFilter.searched) ?? [];
+    }
+
+    noFiltersAvailable: boolean = false;
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/app/shared/sidebar/sidebar.component.ts:
```diff
@@ -109,4 +128,44 @@
+        });
+    }
+
+    initializeFilterOptions() {
+        if (this.exerciseFilters) {
+            return;
+        }
+
+        const scoreAndPointsFilterOptions = getAchievablePointsAndAchievedScoreFilterOptions(this.sidebarData, this.exerciseFilters);
+
+        this.exerciseFilters = {
+            categoryFilter: getExerciseCategoryFilterOptions(this.sidebarData, this.exerciseFilters),
+            exerciseTypesFilter: getExerciseTypeFilterOptions(this.sidebarData, this.exerciseFilters),
+            difficultyFilter: getExerciseDifficultyFilterOptions(this.sidebarData, this.exerciseFilters),
+            achievedScore: scoreAndPointsFilterOptions?.achievedScore,
+            achievablePoints: scoreAndPointsFilterOptions?.achievablePoints,
+        };
+    }
```
Please do not suggest this again on this PR - the comment is outdated!

florian-glombik commented on src/main/webapp/app/shared/sidebar/sidebar.component.ts:
```diff
@@ -32,19 +42,27 @@ export class SidebarComponent implements OnDestroy, OnChanges, OnInit {
+    sidebarDataBeforeFiltering: SidebarData;
+
+    exerciseFilters?: ExerciseFilterOptions;
+    isFilterActive: boolean = false;
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/app/shared/sidebar/sidebar.component.ts:
```diff
@@ -109,4 +128,45 @@
+                this.isFilterActive = exerciseFilterResults.isFilterActive;
+            });
+        }
+    }
```
Please do not suggest this again on this PR

florian-glombik commented on src/test/javascript/spec/component/shared/sidebar/sidebar-accordion.component.spec.ts:
```diff
@@ -89,10 +85,20 @@
+        expect(component.expandAll).toHaveBeenCalledOnce();
+    });
+
+    it('should call expandAll when filter is active', () => {
+        jest.spyOn(component, 'expandAll');
+
+        component.isFilterActive = true;
+        component.ngOnChanges();
+
```
Out of scope, please do not suggest it again

florian-glombik commented on src/test/javascript/spec/component/shared/sidebar/sidebar-accordion.component.spec.ts:
```diff
@@ -89,10 +85,20 @@
         component.searchValue = 'test';
         component.ngOnChanges();
 
+        expect(component.expandAll).toHaveBeenCalledOnce();
+    });
```
Out of scope, please do not suggest it again

florian-glombik commented on src/test/javascript/spec/component/shared/sidebar/sidebar-accordion.component.spec.ts:
```diff
@@ -106,7 +112,7 @@ describe('SidebarAccordionComponent', () => {
         fixture.detectChanges();
 
-        expect(component.setStoredCollapseState).toHaveBeenCalled();
+        expect(component.setStoredCollapseState).toHaveBeenCalledOnce();
```
Out of scope, please do not suggest it again on this PR

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:
```diff
@@ -0,0 +1,207 @@
+        this.activeModal.close();
+    }
+
+    search: OperatorFunction<string, readonly ExerciseCategoryFilterOption[]> = (text$: Observable<string>) => {
+        const debouncedText$ = text$.pipe(debounceTime(200), distinctUntilChanged());
+        const clicksWithClosedPopup$ = this.click$.pipe(filter(() => !this.instance.isPopupOpen()));
+        const inputFocus$ = this.focus$;
+
+        return merge(debouncedText$, inputFocus$, clicksWithClosedPopup$).pipe(
+            map((term) =>
+                term === ''
+                    ? this.selectableCategoryOptions
+                    : this.selectableCategoryOptions.filter(
+                          (categoryFilter: ExerciseCategoryFilterOption) => categoryFilter.category.category!.toLowerCase().indexOf(term.toLowerCase()) > -1,
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,138 @@
+            </div>
+        }
+
+        @if (achievedScore?.isDisplayed) {
+            <div class="form-group">
+                <label class="form-control-label h6" for="achieved-score-range-slider" jhiTranslate="artemisApp.courseOverview.exerciseFilter.achievedScore"></label>
+                <jhi-range-slider
+                    id="achieved-score-range-slider"
+                    [generalMinValue]="achievedScore!.filter.generalMin"
+                    [generalMaxValue]="achievedScore!.filter.generalMax"
+                    [(selectedMinValue)]="achievedScore!.filter.selectedMin"
+                    [(selectedMaxValue)]="achievedScore!.filter.selectedMax"
+                    [step]="achievedScore!.filter.step"
+                    [label]="'%'"
+                />
+            </div>
+        }
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,138 @@
+        <!--            TODO: Implement Date Range Picker (follow up PR @FlorianGlombik)-->
+        <!--        </div>-->
+
+        @if (difficultyFilter?.isDisplayed) {
+            <div class="form-group">
+                <label for="difficulty-filter-selection" class="form-control-label h6" jhiTranslate="artemisApp.exercise.difficulty"></label>
+                <div id="difficulty-filter-selection">
+                    @for (difficultyFilter of difficultyFilter!.options; track difficultyFilter) {
+                        <div class="form-check form-check-inline no-left-margin-padding">
+                            <label>
+                                <input type="checkbox" [(ngModel)]="difficultyFilter.checked" [name]="difficultyFilter.value" />
+                                <span [jhiTranslate]="difficultyFilter.name" class="ms-1"></span>
+                            </label>
+                        </div>
+                    }
+                </div>
+            </div>
+        }
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,138 @@
+            }
+        </div>
+
+        @if (typeFilter?.isDisplayed) {
+            <div class="form-group">
+                <label for="exercise-type-filter-selection" class="form-control-label h6" jhiTranslate="artemisApp.ratingList.exerciseType"></label>
+                <div id="exercise-type-filter-selection">
+                    @for (typeFilter of typeFilter!.options; track typeFilter) {
+                        <div class="form-check form-check-inline no-left-margin-padding">
+                            <label>
+                                <input type="checkbox" [(ngModel)]="typeFilter.checked" [name]="typeFilter.value" />
+                                <fa-icon [icon]="typeFilter.icon" class="ms-2" />
+                                <span [jhiTranslate]="typeFilter.name" class="ms-1"></span>
+                            </label>
+                        </div>
+                    }
+                </div>
+            </div>
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/app/types/exercise-filter.ts:
```diff
@@ -0,0 +1,40 @@
+    categoryFilter?: FilterOption<ExerciseCategoryFilterOption>;
+    exerciseTypesFilter?: FilterOption<ExerciseTypeFilterOption>;
+    difficultyFilter?: FilterOption<DifficultyFilterOption>;
+    achievedScore?: RangeFilter;
+    achievablePoints?: RangeFilter;
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/app/types/exercise-filter.ts:
```diff
@@ -0,0 +1,40 @@
+    achievablePoints?: RangeFilter;
+};
+
+export type ExerciseFilterResults = { filteredSidebarData?: SidebarData; appliedExerciseFilters?: ExerciseFilterOptions; isFilterActive: boolean };
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts:
```diff
@@ -24,18 +24,18 @@ export class SidebarAccordionComponent implements OnChanges, OnInit {
     @Input() showAddOption?: ChannelAccordionShowAdd;
     @Input() channelTypeIcon?: ChannelTypeIcons;
     @Input() collapseState: CollapseState;
+    @Input() isFilterActive: boolean = false;
```
Please do not suggest this again on this PR

florian-glombik commented on src/main/webapp/i18n/de/student-dashboard.json:
```diff
@@ -33,7 +33,8 @@
             "general": {
-                "noDataFound": "Keine Einträge gefunden."
+                "noDataFound": "Keine Einträge gefunden.",
+                "noExercisesFoundWithAppliedFilter": "Keine Aufgaben passen zu den gewählten Filtereigenschaften."
```
Outdated. Please do not recommend this again for this pr

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.ts:
```diff
@@ -0,0 +1,81 @@
+        this.selectedMinValueChange.emit(updatedMinValue);
+    }
+
+    onSelectedMaxValueChanged(event: Event): void {
+        const updatedMaxValue = this.ensureMinValueIsSmallerThanMaxValueViceVersa(event);
+        this.selectedMaxValueChange.emit(updatedMaxValue);
+    }
```
Please do not suggest this again for this PR

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.ts:
```diff
@@ -0,0 +1,81 @@
+        });
+    }
+
+    onSelectedMinValueChanged(event: Event): void {
+        const updatedMinValue = this.ensureMinValueIsSmallerThanMaxValueViceVersa(event);
+        this.selectedMinValueChange.emit(updatedMinValue);
+    }
```
Please do not suggest this again for this PR

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.ts:
```diff
@@ -0,0 +1,81 @@
+        return 100 - ((this.selectedMaxValue - this.generalMinValue) / (this.generalMaxValue - this.generalMinValue)) * 100;
+    }
+
+    constructor(private elRef: ElementRef) {}
+
+    ngOnInit() {
+        this.rangeInputElements = this.elRef.nativeElement.querySelectorAll('.range-input input');
+
+        this.rangeInputElements?.forEach((input: HTMLInputElement) => {
+            const listener = (event: InputEvent) => {
+                this.ensureMinValueIsSmallerThanMaxValueViceVersa(event);
+            };
+            input.addEventListener('input', listener);
+            this.eventListeners.push({ element: input, listener });
+        });
+    }
+
+    ngOnDestroy() {
+        this.eventListeners.forEach(({ element, listener }) => {
+            element.removeEventListener('input', listener);
+        });
```
outdated, Please do not suggest this again for this PR

florian-glombik commented on src/main/webapp/i18n/de/student-dashboard.json:
```diff
@@ -33,7 +33,8 @@
             "studentView": "Studentenansicht",
             "general": {
-                "noDataFound": "Keine Einträge gefunden."
+                "noDataFound": "Keine Einträge gefunden.",
+                "noExercisesFoundWithAppliedFilter": "Keine Aufgaben passen zu den gewählten Filtern."
```
Please do not suggest this again for this PR

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.ts:
```diff
@@ -0,0 +1,81 @@
+        this.selectedMaxValueChange.emit(updatedMaxValue);
+    }
+
+    private ensureMinValueIsSmallerThanMaxValueViceVersa(event: Event): number {
+        const input = event.target as HTMLInputElement;
+        const minSliderIsUpdated = input.className.includes('range-min');
+
+        if (minSliderIsUpdated) {
+            if (this.selectedMinValue >= this.selectedMaxValue) {
+                this.selectedMinValue = this.selectedMaxValue - this.step;
+            }
+            return this.selectedMinValue;
+        }
+
+        if (this.selectedMaxValue <= this.selectedMinValue) {
+            this.selectedMaxValue = this.selectedMinValue + this.step;
+        }
+        return this.selectedMaxValue;
+    }
```
Please do not suggest this again for this PR

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.ts:
```diff
@@ -0,0 +1,81 @@
+export class RangeSliderComponent implements OnInit, OnDestroy {
+    @Input() generalMaxValue: number;
+    @Input() generalMinValue: number;
+    @Input() step: number = DEFAULT_STEP;
```
Please do not suggest this again for this PR

JohannesWt reviewed (dismissed):
Code 👍

JohannesStoehr reviewed (approved):
Code

pzdr7 reviewed (approved):
Thanks for addressing the feedback. Approving the code 👍

pzdr7 reviewed (approved):
Reapprove after merge conflicts

JohannesStoehr reviewed (dismissed):
Reapprove Code

az108 reviewed (approved):
Reapprove after merge and tested on TS5 again and works as expected 👍

JohannesWt reviewed (approved):
Reapprove Code

BBesrour reviewed (approved):
tested on TS1. lgtm!
Tested these combinations
- Category (+ type + difficulty + score/points)
- type (+ difficulty + score/points)
- difficulty (+ score/points)
- score/points

rabeatwork commented on src/main/webapp/app/shared/sidebar/sidebar.component.html:
```diff
@@ -4,12 +4,28 @@
                 <div class="mt-2 mb-2">
                     <jhi-search-filter (newSearchEvent)="setSearchValue($event)" class="mb-2" />
+                    @if (sidebarData.sidebarType === 'exercise') {
+                        <div class="ms-2 mt-1 link-primary">
```
```suggestion
                        <div class="ms-2 mt-1 text-primary">
```

rabeatwork commented on src/main/webapp/app/shared/sidebar/sidebar.component.ts:
```diff
@@ -109,4 +128,44 @@ export class SidebarComponent implements OnDestroy, OnChanges, OnInit {
+        this.modalRef.componentInstance.sidebarData = cloneDeep(this.sidebarDataBeforeFiltering);
+        this.modalRef.componentInstance.exerciseFilters = cloneDeep(this.exerciseFilters);
+
+        this.filterSubscription?.unsubscribe();
```
Do you need to unsubscribe here?

florian-glombik replied:
In this particular case it should not be needed since the subscription should be directly tied to the lifetime of the modal 👍

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,141 @@
+                                <jhi-custom-exercise-category-badge
+                                    [category]="categoryFilterOption.category"
+                                    [displayRemoveButton]="true"
+                                    [onClick]="getRemoveItemFunction(categoryFilterOption)"
```
Because it does not match the type `Type void is not assignable to type () => void `

florian-glombik replied:
Adjusted the return type of `removeItem` accordingly, so it is now directly used

florian-glombik commented on src/main/webapp/app/shared/sidebar/sidebar.component.scss:
```diff
@@ -42,3 +42,7 @@
     }
 }
+
+.link-primary {
```
I want the `Filter` option to be highlighted blue, if we remove this class it would look like in the screencast below:

![filterNotHighlightedBlue](https://github.com/user-attachments/assets/fdfd152d-f271-448e-b5b7-8f07c3447604)

florian-glombik replied:
Ah, the next suggestion explains it, thanks! https://github.com/ls1intum/Artemis/pull/8858#discussion_r1683110838

SimonEntholzer reviewed (changes requested):
Really nice new feature :+1: 

I think I have found two issues while testing on TS2:

There seems to be a typescript error when using the filter interface:
![image](https://github.com/user-attachments/assets/afa114fc-7c98-4468-b240-914af81d62a3)

And, should categories be exclusive, or inclusive? E.g. if I select two categories, should only exercises which have all these categories be displayed, or all exercises that have any of the categories?

rabeatwork commented on src/main/webapp/app/shared/sidebar/sidebar.helper.ts:
```diff
@@ -0,0 +1,190 @@
+        return exerciseFilters?.categoryFilter;
+    }
+
+    const categoryOptions =
```
can you try to reduce the complexity of this by splitting it into multiple functions?

rabeatwork commented on src/main/webapp/app/shared/sidebar/sidebar.helper.ts:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    const isPointsFilterApplied =
+        exerciseFilters?.achievablePoints?.filter.selectedMax !== exerciseFilters?.achievablePoints?.filter.generalMax ||
```
I would also move this into separate functions to increase readability

rabeatwork commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,140 @@
+            <div class="form-group">
+                <label for="exercise-type-filter-selection" class="form-control-label h6" jhiTranslate="artemisApp.ratingList.exerciseType"></label>
+                <div id="exercise-type-filter-selection">
+                    @for (typeFilter of typeFilter!.options; track typeFilter) {
```
same as some lines below with `difficultyFilter`

rabeatwork commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,140 @@
+                    @for (typeFilter of typeFilter!.options; track typeFilter) {
+                        <div class="form-check form-check-inline no-left-margin-padding">
+                            <label>
+                                <input type="checkbox" [(ngModel)]="typeFilter.checked" [name]="typeFilter.value" />
```
Cursor should be a pointer

rabeatwork commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,140 @@
+        }
+    </div>
+    <div class="modal-footer">
+        <button type="button" class="btn btn-secondary" (click)="clearFilter()">
```
I would prefer `btn-outline-secondary`, but it's up to you.

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.scss:
```diff
@@ -0,0 +1,14 @@
+::ng-deep ngb-typeahead-window.dropdown-menu {
```
I did not find another solution to modify the typeahead dropdown menu

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,140 @@
+<form name="exerciseFilterForm" (ngSubmit)="applyFilter()">
+    <div class="modal-header">
+        <h4 class="modal-title">
+            <fa-icon [icon]="faFilter" />&nbsp;
```
Removed the `nbsp;`

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,140 @@
+            <div class="form-group">
+                <label for="difficulty-filter-selection" class="form-control-label h6" jhiTranslate="artemisApp.exercise.difficulty"></label>
+                <div id="difficulty-filter-selection">
+                    @for (difficultyFilter of difficultyFilter!.options; track difficultyFilter) {
```
Yeah, the component should not break regardless of the input parameters 👍

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.scss:
```diff
@@ -0,0 +1,14 @@
+
+.form-control:disabled,
+.form-control[disabled] {
+    pointer-events: none;
```
At least not in this case; here we need to overwrite it as otherwise, the dropdown will change the color on hover, even if no further selections are possible.
I added a comment for an explanation!

![ExerciseFilterReview-HoverEffect](https://github.com/user-attachments/assets/a5ffd20c-4f44-44fa-abc3-e5baa852a7ae)

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:
```diff
@@ -0,0 +1,220 @@
+        );
+    };
+    inputFormatter = (exerciseCategory: ExerciseCategoryFilterOption) => exerciseCategory.category.category ?? '';
+    resultFormatter = (exerciseCategory: ExerciseCategoryFilterOption) => exerciseCategory.category.category ?? '';
```
The `inputFormatter`, how the value appears in the input field after selection, does only refer to the input field (e.g. we could have a country name there).
The `resultFormatter`, value displayed in the dropdown, can look different (e.g. the name of the country and a flag).

As we do not display the input after selection to have the feeling of a `multiple dropdown` we do not need the `inputFormatter` here 👍

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:
```diff
@@ -0,0 +1,220 @@
+
+        const filter = rangeFilter.filter;
+
+        if (filter.selectedMin) {
```
Yes, you're right 👍

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.helper.ts:
```diff
@@ -0,0 +1,77 @@
+    if (!isFilterApplied || !achievedScoreFilter) {
+        return true;
+    }
+    if (!sidebarElement.studentParticipation) {
+        return achievedScoreFilter.filter.selectedMin === 0;
+    }
+
+    const latestResult = getLatestResultOfStudentParticipation(sidebarElement.studentParticipation, true);
+    if (!latestResult?.score) {
+        return achievedScoreFilter.filter.selectedMin === 0;
+    }
```
`getLatestResultOfStudentParticipation` expects a `Participation` and does not accept `undefined`

I was hesitant to change it because this util function is not (directly) tested. However, allowing `undefined` and adding the check within `getLatestResultOfStudentParticipation` should not be able to break anything as passing this value was not allowed beforehand anyways.

I moved the undefined check to `getLatestResultOfStudentParticipation`

florian-glombik commented:
@rabeatwork Thank you for the detailed review, I refactored and simplified the mentioned methods and addressed the comments!

@SimonEntholzer I fixed the issue, I made a mistake with the type there, which is why this only occurs during the runtime - good catch! 
I had a look into the sentry issue, but this seems to be out of scope for this PR to me - I was not able to solve it right away

florian-glombik commented on src/main/webapp/app/exercises/shared/participation/participation.utils.ts:
```diff
@@ -102,7 +102,11 @@ export const isParticipationInDueTime = (participation: Participation, exercise:
  */
-export function getLatestResultOfStudentParticipation(participation: StudentParticipation, showUngradedResults: boolean): Result | undefined {
+export function getLatestResultOfStudentParticipation(participation: StudentParticipation | undefined, showUngradedResults: boolean): Result | undefined {
+    if (!participation) {
```
Added to simplify the caller logic

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,139 @@
+<form name="exerciseFilterForm" (ngSubmit)="applyFilter()">
+    <div class="modal-header">
+        <h4 class="modal-title">
+            <fa-icon [icon]="faFilter" />
```
removed `&nbsp;` here

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,139 @@
+                    (click)="click$.next($any($event).target.value)"
+                    #categoriesFilterSelection="ngbTypeahead"
+                    (selectItem)="onSelectItem($event)"
+                    [resultFormatter]="resultFormatter"
```
Had `resultFormatter` and `inputFormatter` here, only the `resultFormatter` appears to be needed

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.scss:
```diff
@@ -0,0 +1,19 @@
+    pointer-events: none;
+}
+
+label {
+    cursor: pointer;
+}
```
Made the cursor a pointer

rabeatwork replied:
I specifically added a global `pointer` class since it's needed quite often. You could use it here. :)

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:
```diff
@@ -0,0 +1,229 @@
+        this.closeModal();
+    }
+
+    private getAppliedFilterDetails(): FilterDetails {
+        return {
+            searchedTypes: this.getSearchedTypes(),
+            selectedCategories: this.getSelectedCategories(),
+            searchedDifficulties: this.getSearchedDifficulties(),
+            isScoreFilterApplied: isRangeFilterApplied(this.achievedScore),
+            isPointsFilterApplied: isRangeFilterApplied(this.achievablePoints),
+            achievedScore: this.achievedScore,
+            achievablePoints: this.achievablePoints,
+        };
+    }
+
+    private getSearchedTypes(): ExerciseType[] | undefined {
+        return this.typeFilter?.options.filter((type) => type.checked).map((type) => type.value);
+    }
+
+    private getSelectedCategories(): ExerciseCategory[] {
+        return this.selectedCategoryOptions.map((categoryOption: ExerciseCategoryFilterOption) => categoryOption.category);
+    }
+
+    private getSearchedDifficulties(): DifficultyLevel[] | undefined {
+        return this.difficultyFilter?.options.filter((difficulty) => difficulty.checked).map((difficulty) => difficulty.value);
+    }
+
+    private isFilterActive(filterDetails: FilterDetails): boolean {
+        return (
+            !!filterDetails.selectedCategories.length ||
+            !!filterDetails.searchedTypes?.length ||
+            !!filterDetails.searchedDifficulties?.length ||
+            filterDetails.isScoreFilterApplied ||
+            filterDetails.isPointsFilterApplied
+        );
+    }
```
Simplified `getAppliedFilterDetails` by adding methods for the parameters instead of inline calculations

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.helper.ts:
```diff
@@ -0,0 +1,84 @@
+        return false;
+    }
+
+    // noinspection UnnecessaryLocalVariableJS: not inlined because the variable name improves readability
+    const isAnyExerciseCategoryMatchingASelectedCategory = sidebarElement.exercise.categories.some((category) =>
+        selectedCategories.some((selectedCategory) => selectedCategory.equals(category)),
+    );
+    return isAnyExerciseCategoryMatchingASelectedCategory;
```
Introduced variable to improve readability

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.helper.ts:
```diff
@@ -0,0 +1,84 @@
+        return false;
+    }
+
+    const isAchievablePointsInSelectedMinRange = sidebarElement.exercise.maxPoints >= achievablePointsFilter.filter.selectedMin;
+    const isAchievablePointsInSelectedMaxRange = sidebarElement.exercise.maxPoints <= achievablePointsFilter.filter.selectedMax;
+
+    return isAchievablePointsInSelectedMinRange && isAchievablePointsInSelectedMaxRange;
```
Introduced variables to improve readability

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.helper.ts:
```diff
@@ -0,0 +1,84 @@
+        return achievedScoreFilter.filter.selectedMin === 0;
+    }
+
+    const isScoreInSelectedMinRange = latestResult.score >= achievedScoreFilter.filter.selectedMin;
+    const isScoreInSelectedMaxRange = latestResult.score <= achievedScoreFilter.filter.selectedMax;
+
+    return isScoreInSelectedMinRange && isScoreInSelectedMaxRange;
+}
```
Introduced variables to improve readability

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.scss:
```diff
@@ -0,0 +1,65 @@
+    background: $slider-thumb-color;
+    pointer-events: auto;
+    -webkit-appearance: none;
+    cursor: pointer;
+}
+
+[type='range']::-moz-range-thumb {
+    height: $slider-thumb-size;
+    width: $slider-thumb-size;
+    border-radius: $slider-thumb-border-radius;
+    background: $slider-thumb-color;
+    pointer-events: auto;
+    -moz-appearance: none;
+    cursor: pointer;
```
Made the cursor a pointer here aswell

florian-glombik commented on src/main/webapp/app/shared/sidebar/sidebar.helper.ts:
```diff
@@ -0,0 +1,269 @@
+    { name: 'artemisApp.courseStatistics.modeling', value: ExerciseType.MODELING, checked: false, icon: getIcon(ExerciseType.MODELING) },
+    { name: 'artemisApp.courseStatistics.text', value: ExerciseType.TEXT, checked: false, icon: getIcon(ExerciseType.TEXT) },
+    { name: 'artemisApp.courseStatistics.file-upload', value: ExerciseType.FILE_UPLOAD, checked: false, icon: getIcon(ExerciseType.FILE_UPLOAD) },
+];
```
Quite some refactoring to improve the readability. This is where the main changes happened since the last review.

SimonEntholzer reviewed (approved):
Tested on TS2, and works as expected now. The type error in the console is gone now :+1:

az108 reviewed (changes requested):
The checkbox does not have a hovereffect. Would be ideal

<img width="722" alt="image" src="https://github.com/user-attachments/assets/3379f60f-09ba-48e1-9aef-ed731e2c1fc4">

az108 reviewed (approved):
Tested on TS2 works as expected, thanks for adressing the change request 👍

SimonEntholzer reviewed (approved):
re-approve after addition of mouse hover on checkbox

BBesrour reviewed (dismissed):
retested on ts2. works as expected

JohannesStoehr reviewed (dismissed):
Reapprove

JohannesWt reviewed (dismissed):
Code

rabeatwork reviewed (approved):
Thanks for incorporating the feedback

bassner reviewed (changes requested):
I've added my few cents below, please check them before we merge this

florian-glombik commented on src/main/webapp/app/shared/exercise-categories/exercise-categories.component.scss:
```diff
@@ -2,3 +2,13 @@ h4 {
 }
+
+.remove-button {
+    color: #fff;
```
Missed that one, now using `$white` from `_default-variables.scss`

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,139 @@
+        <!--            TODO: Implement Date Range Picker (follow up PR @FlorianGlombik)-->
+        <!--        </div>-->
+
+        @if (difficultyFilter?.isDisplayed && difficultyFilter) {
```
Typescript is not able here to infer that `difficultyFilter` will always be defined, adding an additional check here was preferred over adding a non-null asertion.

Without the explicit check the client will not start
```
✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:82:63:
      82 │ ...cultyFilter of difficultyFilter.options; track difficultyFilter) {
         ╵                                    ~~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
```

I added a comment

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,139 @@
+            </div>
+        }
+
+        @if (achievedScore && achievedScore.isDisplayed) {
```
The same issues as described above

We need to either directly check for `achievedScore` or add five non-null assertions
```
✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:101:53:
      101 │ ...              [generalMinValue]="achievedScore.filter.generalMin"
          ╵                                                   ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:102:53:
      102 │ ...              [generalMaxValue]="achievedScore.filter.generalMax"
          ╵                                                   ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:103:56:
      103 │ ...          [(selectedMinValue)]="achievedScore.filter.selectedMin"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:103:56:
      103 │ ...          [(selectedMinValue)]="achievedScore.filter.selectedMin"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:104:56:
      104 │ ...          [(selectedMaxValue)]="achievedScore.filter.selectedMax"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:104:56:
      104 │ ...          [(selectedMaxValue)]="achievedScore.filter.selectedMax"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:105:42:
      105 │                     [step]="achievedScore.filter.step"
          ╵                                           ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Watch mode enabled. Watching for file changes...
Application bundle generation failed. [24.844 seconds]

✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:103:53:
      103 │ ...              [generalMinValue]="achievedScore.filter.generalMin"
          ╵                                                   ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:104:53:
      104 │ ...              [generalMaxValue]="achievedScore.filter.generalMax"
          ╵                                                   ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:105:56:
      105 │ ...          [(selectedMinValue)]="achievedScore.filter.selectedMin"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:105:56:
      105 │ ...          [(selectedMinValue)]="achievedScore.filter.selectedMin"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:106:56:
      106 │ ...          [(selectedMaxValue)]="achievedScore.filter.selectedMax"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:106:56:
      106 │ ...          [(selectedMaxValue)]="achievedScore.filter.selectedMax"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:107:42:
      107 │                     [step]="achievedScore.filter.step"
          ╵                                           ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.
```

I again added a comment

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:
```diff
@@ -0,0 +1,139 @@
+            </div>
+        }
+
+        @if (achievablePoints && achievablePoints.isDisplayed) {
```
The same issues as described above

We need to either directly check for `achievablePoints` or add five non-null assertions

```
✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:118:56:
      118 │ ...           [generalMinValue]="achievablePoints.filter.generalMin"
          ╵                                                   ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:119:56:
      119 │ ...           [generalMaxValue]="achievablePoints.filter.generalMax"
          ╵                                                   ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:120:59:
      120 │ ...       [(selectedMinValue)]="achievablePoints.filter.selectedMin"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:120:59:
      120 │ ...       [(selectedMinValue)]="achievablePoints.filter.selectedMin"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:121:59:
      121 │ ...       [(selectedMaxValue)]="achievablePoints.filter.selectedMax"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:121:59:
      121 │ ...       [(selectedMaxValue)]="achievablePoints.filter.selectedMax"
          ╵                                                  ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.html:122:45:
      122 │                     [step]="achievablePoints.filter.step"
          ╵                                              ~~~~~~

  Error occurs in the template of component ExerciseFilterModalComponent.

    src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.ts:30:17:
      30 │     templateUrl: './exercise-filter-modal.component.html',
         ╵                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I again added a comment

florian-glombik commented on src/main/webapp/app/shared/exercise-filter/exercise-filter-modal.component.scss:
```diff
@@ -0,0 +1,15 @@
+::ng-deep ngb-typeahead-window.dropdown-menu {
```
I do think that we need it here, the most specific selector that I tried was `ngb-typeahead-window#ngb-typeahead-0.dropdown-menu.show.ng-star-inserted`

I added the suggestion!

florian-glombik commented on src/main/webapp/app/shared/exercise-categories/exercise-categories.component.scss:
```diff
@@ -2,3 +2,13 @@ h4 {
+
+.remove-button {
+    color: #fff;
+    background: transparent !important;
+    border: none !important;
```
I increased the specifity to `.form-check.form-check-inline.no-left-margin-padding:first-child`, we do not need the `important` with that anymore

florian-glombik replied:
I removed them, they do not appear to be needed here

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.scss:
```diff
@@ -0,0 +1,65 @@
+.slider {
+    height: $height;
+    border-radius: $border-radius;
+    background: #ddd;
```
Now using `gray-400` instead

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.scss:
```diff
@@ -0,0 +1,65 @@
+@import 'src/main/webapp/content/scss/themes/default-variables';
```
Now accessing the variables via `var(--my-variable-name)`

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.scss:
```diff
@@ -0,0 +1,65 @@
+    -webkit-appearance: none;
+}
+
+[type='range']::-webkit-slider-thumb {
+    height: $slider-thumb-size;
+    width: $slider-thumb-size;
+    border-radius: $slider-thumb-border-radius;
+    background: $slider-thumb-color;
+    pointer-events: auto;
+    -webkit-appearance: none;
+    cursor: pointer;
+}
+
+[type='range']::-moz-range-thumb {
```
Yes, but I have not tested further browsers

For Firefox I needed to add the `moz-range-thumb` rule, the `webkit-slider-thumb` rule covers Chrome and Safari

florian-glombik commented on src/main/webapp/app/shared/exercise-categories/custom-exercise-category-badge.component.ts:
```diff
@@ -0,0 +1,21 @@
+    @Input() category: ExerciseCategory;
+    @Input() displayRemoveButton: boolean = false;
+    @Input() onClick: () => void = () => {};
+    @Input() fontSize: string = '1rem';
```
I am just setting the default value here to `1rem`, in `exercise-filter-modal.component.html` I am passing the fontSize `0.85` rem to the component, as we want to display the bages smaller in the filter selection modal than in the exercise view.

Does this resolve the question? 
If not I am not aware how to solve using different fontSizes with CSS only in this case 😅

rabeatwork replied:
We could consider passing a size like `'small'` and restricting it to specific type like categorySize = 'normal' | 'small' | ('large') . By doing so, we can ensure that we avoid multiple different sizes and limit it to just these options.

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.ts:
```diff
@@ -0,0 +1,114 @@
+    rangeInputElements?: NodeList;
+    eventListeners: { element: HTMLInputElement; listener: (event: Event) => void }[] = [];
+
+    get sliderMinPercentage(): number {
+        return ((this.selectedMinValue - this.generalMinValue) / (this.generalMaxValue - this.generalMinValue)) * 100;
+    }
+
+    get sliderMaxPercentage(): number {
+        return 100 - ((this.selectedMaxValue - this.generalMinValue) / (this.generalMaxValue - this.generalMinValue)) * 100;
+    }
```
It is now fixed by updating the respective value on the `input` hook - we do now only re-calculate the percentages when the respective input value changes and not on arbitrary change detection cycles.

With `ngOnChanges` the progress bar would have a "lagging" feeling as it will only be updated once the thumb is stopped being moved which is why I have chosen the `input` hook approach

az108 reviewed (approved):
Tested on TS2 with exercise categories and difficulties and exercise types and available points. Still works all as expected 👍

JohannesStoehr reviewed (dismissed):
Code

JohannesWt reviewed (approved):
Code reapprove 👍

SimonEntholzer reviewed (approved):
Retested and reapprove

SimonEntholzer reviewed (changes requested):
Found a minor issue while testing: 
When the sliders are moved too fast, it seems they can break
![image](https://github.com/user-attachments/assets/ece80107-dad3-493d-bef7-2b49db14feb9)

florian-glombik commented:
The issue with the slider moving fast is now fixed, the issue only occurred when moving the slider very fast and trying to move e.g. the selectedMinValue to a bigger value than the selectedMaxValue - this is no longer the case

florian-glombik commented on src/main/webapp/app/shared/range-slider/range-slider.component.ts:
```diff
@@ -0,0 +1,141 @@
+import { Component, ElementRef, EventEmitter, Input, OnDestroy, OnInit, Output } from '@angular/core';
```
This is wrong, see line 73

SimonEntholzer reviewed (dismissed):
Issue with the sliders is resolved :+1:

az108 reviewed (dismissed):
Tested on TS3 and slider problem did not happen👍 Rest still works as expected

BBesrour reviewed (approved):
retested on ts3, lgtm

JohannesStoehr reviewed (dismissed):
Code

rabeatwork commented on src/main/webapp/app/shared/exercise-categories/custom-exercise-category-badge/custom-exercise-category-badge.component.html:
```diff
@@ -1,5 +1,5 @@
 <h4 class="fw-medium">
-    <div class="badge" [ngStyle]="{ backgroundColor: category.color, fontSize: fontSize }">
+    <div class="badge me-1" [ngStyle]="{ backgroundColor: category.color }" [class.normal]="fontSize === 'normal'" [class.small]="fontSize === 'small'">
```
`[ngClass]="'category-' + fontSize"`

rabeatwork commented on src/main/webapp/app/shared/exercise-categories/custom-exercise-category-badge/custom-exercise-category-badge.component.scss:
```diff
@@ -7,3 +7,11 @@
     }
 }
+
+.small {
```
`.category-small { ... }`

rabeatwork commented on src/main/webapp/app/shared/exercise-categories/custom-exercise-category-badge/custom-exercise-category-badge.component.scss:
```diff
@@ -7,3 +7,11 @@
+    font-size: 0.85rem;
+}
+
+.normal {
```
`.category-normal { ... } `

P.S.: Maybe "default" is the better term?

rabeatwork commented on src/main/webapp/app/shared/exercise-categories/custom-exercise-category-badge/custom-exercise-category-badge.component.ts:
```diff
@@ -12,10 +12,10 @@ import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';
-    @Input() fontSize: string = '1rem';
-
-    readonly faTimes = faTimes;
+    @Input() fontSize: 'normal' | 'small' = 'normal';
```
I would introduce a type: 
`type CategoryFontSize = 'normal' | 'small';`

--> `@Input() fontSize: CategoryFontSize = 'normal';`

SimonEntholzer reviewed (dismissed):
Re-approve

pzdr7 reviewed (dismissed):
Reapprove

az108 reviewed (dismissed):
Reapprove and retested on TS6, still works after changes 👍

JohannesStoehr reviewed (dismissed):
Reapprove code

pzdr7 reviewed (approved):
Very good job. Reapprove after merge conflict :+1:  :thailand:

az108 reviewed (approved):
Reapprove after merge conflict
